File: scripts/dep.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import os
import argparse

class VerilogFile:
    #TODO
    @staticmethod
    def read_ignore_list(self, fname):
        VerilogFile.ignore_file = []
        VerilogFile.ignore_com = []
        data_dict = {}
        data = open(fname, 'r').read()
        data = re.sub(r'#.*$', '', data)
        data = re.sub(r'\s+', ' ', data, re.I | re.S)
        for i in re.findall(r'(\S+)\s*\{(.*?)\}'):
            typ, d = i
            d = d.strip().split()
            if typ == 'ignore_file':
                VerilogFile.ignore_file = d
            elif typ == 'ignore_com':
                VerilogFile.ignore_com = [ i+'.com' for i in d ]
            else:
                raise

    def __init__(self, module_name):
        self._module_name = module_name
        self.filename = module_name + '.v'
        #self.filename_body = self.filename.split('.')[0]
        #self.filename_ext = '.'.join(self.filename.split('.')[1:])
        self.filename_body, self.filename_ext = re.findall(r'^([^.]*)\.(.*)$')[0]
        self.dep = None

    def have_v_in(self):
        if self.have_v_erb():
            return True
        else:
            return os.path.exists(self.v_in())

    def have_v_erb(self):
        return os.path.exists(self.v_erb())

    def v_in(self):
        return self.filename_body + '.v.in'
    def v_erb(self):
        return self.filename_body + '.v.erb'
    def com(self):
        return self.filename_body + '.com'

    def walk(self, files):
        text = ''
        if self.have_v_in():
            if not len(self.dep) != 0:
                text += '{}: {}\n'.format(self.filename, ' '.join(self.dep))
        files.append(self.filename)
        if self.have_v_erb():
            text += '{}: {}\n'.format(self.v_in(), self.v_erb())
            files.append(self.v_in())
        if not any([
            self.com() in self.ignore_com,
            len(self.com_dep()) != 0,
            '<' in self.com()
        ]):
            text += '{}: {}\n'.format(self.com(), ' '.join(self.com_dep()))
            files.append(self.com())
            files += [h for h in self.com_dep() if h.endswith('.vh')]
            for i in self.submodules():
                if not i+'.v' in files:
                    if i in self.ignore_file or '<' in i:
                        continue
                    else:
                    files.append(i+'.v')
                    text += VerilogFile(i).walk(files)

        return text

    def com_dep(self):
        text = self.includes() + [ VerilogFile(i).com() for i in submodules]
        text = [ i for i in text if not i in self.ignore_com and not '<' in i ]
        return text

    def dep(self):
        if self.dep != None:
            return self.dep
        self.dep = []
        self.dep += [ i+'v' for i in self.submodules() if not '<' in i ]
        return self.dep

    def data(self):
        if self.data != None:
            return self.data
        if self.have_v_erb():
            self.data = open(self.v_erb, 'r').read()
        elif self.have_v_in():
            self.data = open(self.v_in, 'r').read()
        else:
            if os.path.exists(self.filename()):
                self.data = open(self.filename(), 'r').read()
            else:
                self.data = ''
        return self.data

    def _submodules(self, string):
    v_files = []
    data0 = [ i[0] for i in re.findall(r'([a-zA-Z_0-9]+)\s+\S+\s*\((?:\s*\.\S+\s*\([^\)]*\)\s*,{0,1}){0,}(?:\s*\/\*AUTOINST\*\/\s*){1}\);', string) ]
    v_files += data0
    data1 = [ i[0] for i in re.findall(r'([a-zA-Z_0-9]+)\s+\S+\s*\((?:\s*\.\S+\s*\([^\)]*\)\s*,{0,1}){1,}\);', string) ]
    v_files += data1
    return list(set(v_files))

    #TODO
    def submodules(self):
        v_files = self._submodules(data)
        sdata = #TODO
        v_files += self._submodules(sdata)
        text_list = v_files

        return  sorted(text_list, key=str.lower)

    #TODO
    def includes(self):
    v_files = [ [i[0]] + VHeaderFile(i[0]).includes()
                    for i in re.findall(r'^\s*`include\s+"([^"]*)"\s*$') ]

    return 

    #TODO
    def to_dot_sub(self):
        text = []
        text.append('node_{} [label=\"{}\"];'.format(self.filename_body, filename_body))
        text += [ 

    def to_dot(self):
        return 'digraph sample {{\n' \
               '{}\n'.format(self.to_dot_sub()) \
               '}}\n' \

class VHeaderFile(VerilogFile):
    def __init__(self, filename):
        self.filename = filename
        self.filename_body = self.filename.split('.')[0]
        self.filename_ext = '.'.join(self.filename.split('.')[1:])
        self.dep = []

class VFiles(list):
    def to_str(self):
        f = list(set(self))
        f = self.purge(f)
        f = r' \\\n'.join(sorted(f, key=str.lower)) + '\n'
        return r'{}= \\\n{}\n'.format(self.name(), f)
    def name(self):
        return 'VFILES'
    def purge(self, f):
        f = [ i for i in f if i.endswith('.v') ]
        f = [ i for i in f if not i.startswith('alt') ]
        return

class VHFiles(VFiles):
    def name(self):
        return 'VHFILES'
    def purge(self, f):
        f = [ i for i in f if i.endswith('.vh') ]
        f = [ i for i in f if not i.startswith('alt') ]

parser = argparse.ArgumentParser(
    description='Verilog-mode Source Parser')
parser.add_argument('--dc', help='design compiler readfile script')
parser.add_argument('--dot', help='generate dot file')
parser.add_argument('--make', help='generate Makefile (default)')
parser.add_argument('files', help='verilog source files')
args = parser.parse_args()

VerilogFile.read_ignore_list('dep.in')

top_v_files = [ VFile(top) for top in args.files ]

#TODO
top_vfiles_str =

files = files_all

if args.make:
    print('## This Makefile is generated by dep.py\n')
    print(VFiles(files).to_str())
    print(VHFiles(files).to_str())
    print('\n'.join(top_vfiles_str))
elif args.dc:
    files = [ i for i in files if i.endswith('.v') ]
    for i in list(set(files)):
        print('read_file -format verilog {}\n'.format(i))
elif args.dot:
    for t in top_vfiles:
        print t.to_dot()

File: rtl/dep.rb
#!/usr/local/bin/ruby

require 'optparse'
require 'ostruct'
require 'pp'

class VFile
    attr_reader(:filename)
    attr_reader(:module_name)
    class << self
        def read_ignore_list(file)
            @@ignore_file=[]
            @@ignore_com=[]
            data_hash={}
            data=open(file).read
            data=data.gsub(/#.*$/,"")
            data=data.gsub(/\s+/im," ")
            data.scan(/(\S+)\s*\{(.*?)\}/).each do |i|
                type,d=i
                d=d.strip.split(/\s+/)
                case type
                when "ignore_file"
                    @@ignore_file=d
                when "ignore_com"
                    @@ignore_com=d.map{|i| "#{i}.com"}
                else
                    raise
                end
            end
        end
    end
    def initialize(module_name)
        @module_name=module_name
        @filename=module_name+".v"
        @filename_body,@filename_ext=@filename.scan(/^([^.]*)\.(.*)$/)[0]
        @dep=nil
    end
    def have_v_in
        if have_v_erb
            true
        else
            File.exist?(v_in)
        end
    end
    def have_v_erb
        File.exist?(v_erb)
    end
    def v_in;    dep_candidate=@filename_body+".v.in"    end
    def v_erb; dep_candidate=@filename_body+".v.erb" end
    def com;     dep_candidate=@filename_body+".com"     end
    def walk(files)
        r=""
        if have_v_in
            if not dep.empty?
                r+=<<EOF
#{@filename}: #{dep*" "}
EOF
            end
        end
        files<< @filename
        if have_v_erb
            r+=<<EOF
#{v_in}: #{v_erb}
EOF
            files<< v_in
        end
#        pp com
        if not @@ignore_com.include?(com) and
                not (c=com_dep).empty? and
                not com=~/</
            r+=<<EOF
#{com}: #{c*" "}
EOF
            files<< com
            c.each do |i|
                if i=~/\.vh$/
                    files<< i
                end
            end
        end
        submodules.each do |i|
            if not files.include?(i+".v")
                next if @@ignore_file.include?(i) or i=~/</
                files<< i+".v"
                r+=self.class.new(i).walk(files)
            end
        end
        return r
    end
    def com_dep
        r=includes+submodules.map{|i| self.class.new(i).com}
        r.delete_if{|i| @@ignore_com.include?(i) or i=~/</ }
        return r
    end
    def dep
        return @dep if @dep!=nil
        @dep=[]
#        @dep<< v_in if have_v_in
        s=submodules
        @dep+=s.map{|i|
            i+".v"
        }
        @dep.delete_if{|i| i=~/</ }
        @dep
    end
    def data
        return @data if @data!=nil
        if have_v_erb
            @data=open(v_erb).read
        elsif have_v_in
            @data=open(v_in).read
        else
            if File.exist?(filename)
                @data=open(filename).read
            else
                @data=""
            end
        end
        return @data
    end
    def _submodules(str)
        v_files=[]
        data0=str.scan(/([a-zA-Z_0-9]+)\s+\S+\s*\((?:\s*\.\S+\s*\([^\)]*\)\s*,{0,1}){0,}(?:\s*\/\*AUTOINST\*\/\s*){1}\);/).map do |i|
            i[0]
        end
        v_files+=data0
        data1=str.scan(/([a-zA-Z_0-9]+)\s+\S+\s*\((?:\s*\.\S+\s*\([^\)]*\)\s*,{0,1}){1,}\);/).map do |i|
            i[0]
        end
        v_files+=data1
        return v_files.uniq
    end
    def submodules
        v_files=_submodules(data)
        sdata=data.gsub(/\/\/.*$/,"").gsub(/\/\*.*?\*\//,"").gsub(/\s+/," ")
        v_files+=_submodules(sdata)
        r=v_files
#### hack
        if @filename=~/^sramw_\d+_(\d+_\d+)/
            r+=["ioconv_sram_#{$1}","sram_2_#{$1}","sramip_2_#{$1}"]
        elsif @filename=~/^sramw_(\d+_\d+)/
            r+=["ioconv_sram_#{$1}","sram_#{$1}","sramip_#{$1}"]
        end
#### end of hack
        r=r.uniq.sort
        return r
    end
    def includes
        v_files=data.scan(/^\s*`include\s+"([^"]*)"\s*$/).map do |i|
            [i[0]]+VHFile.new(i[0]).includes
        end
        return v_files.flatten.uniq
    end
    def to_dot_sub
        r=[]
        r<< "node_#{@filename_body} [label=\"#{@filename_body}\"];"
        r+=submodules.map{|i|
            next if i=~/</

            k=self.class.new(i).to_dot_sub
            r<< "node_#{@filename_body}->node_#{i};"
            k
        }
        r.flatten*"\n"
    end
    def to_dot
        <<EOF
digraph sample {
#{to_dot_sub}
}
EOF
    end
end

class VHFile < VFile
    def initialize(filename)
        @filename=filename # +".v"
        @filename_body,@filename_ext=@filename.scan(/^([^.]*)\.(.*)$/)[0]
        @dep=nil
    end
end

options=OpenStruct.new
options.mode=:make
opts=OptionParser.new do |opts|
    opts.on("--dc","design compiler readfile script") do |i|
        options.mode=:dc
    end
    opts.on("--dot","generate dot file") do |i|
        options.mode=:dot
    end
    opts.on("--make","generate Makefile (default)") do |i|
        options.mode=:make
    end
    opts.on_tail("-h", "--help", "Show this message") do
        puts opts
        exit
    end
end
opts.parse!(ARGV)

VFile.read_ignore_list("dep.in")

files_all=[]
top_vfiles=ARGV.map{|top|
    t=VFile.new(top)
}
top_vfiles_s=top_vfiles.map{|t|
    files=[]
    r=t.walk(files)
    files_all+=files

    f=files.uniq
    f.delete_if{|i| not i=~/\.v$/}
    f.delete_if{|i| i=~/^alt/}
    k=t.module_name
    r+="\n"
    r+="#{k}_FILES= \\\n"
    r+="#{f*" \\\n"}\n"
    r+="\n"
    b=[]
    b<< "#{k}: $(#{k}_FILES) $(#{k}_FILES_OPT)"
    b<< "ifeq ($(SIMULATOR),vcs)"
    b<< "\t$(VCS) -o $@ $(#{k}_FILES) $(#{k}_FILES_OPT)\n"
    b<< "else"
    b<< "ifeq ($(SIMULATOR),iverilog)"
    b<< "\t$(IVERILOG) -o $@ $(#{k}_FILES) $(#{k}_FILES_OPT)\n"
    b<< "endif"
    b<< "endif"
    r+=b.map{|i| i+"\n"}*""
    r
}
files=files_all

class VFiles < Array
    def to_s
        f=self.uniq
        f=purge(f)
        f=f.sort*" \\\n"+"\n"
        "#{name}= \\\n#{f}\n"
    end
    def name
        "VFILES"
    end
    def purge(f)
        f.delete_if{|i| not i=~/\.v$/}
        f.delete_if{|i| i=~/^alt/}
    end
end
class VHFiles < VFiles
    def name
        "VHFILES"
    end
    def purge(f)
        f.delete_if{|i| not i=~/\.vh$/}
        f.delete_if{|i| i=~/^alt/}
    end
end

if options.mode==:make
    print "## -*- makefile -*-\n"
    print VFiles.new(files.dup).to_s
    print VHFiles.new(files.dup).to_s
    print top_vfiles_s*"\n"
elsif options.mode==:dc
    files.delete_if{|i| not i=~/\.v$/}
    print files.uniq.map{|i| "read_file -format verilog #{i}\n"}*""
elsif options.mode==:dot
    top_vfiles.each do |t|
        print t.to_dot
    end
end